pika/connection.py

```diff
@@ -1159,35 +1163,22 @@ def connect(self):
         Connection object should connect on its own.
 
         """
+        assert self._connection_attempt_timer is None, (
+            'connect timer was already scheduled')
+
+        assert self.is_closed, (
+            'connect expected CLOSED state, but got: {}'.format(
+                self._STATE_NAMES[self.connection_state]))
+
         self._set_connection_state(self.CONNECTION_INIT)
 
-        error = self._adapter_connect()
-        if not error:
-            return self._on_connected()
-        self.remaining_connection_attempts -= 1
-        LOGGER.warning('Could not connect, %i attempts left',
-                       self.remaining_connection_attempts)
-        if self.remaining_connection_attempts > 0:
-            LOGGER.info('Retrying in %i seconds', self.params.retry_delay)
-            # TODO remove timeout if connection is closed before timer fires
-            self.add_timeout(self.params.retry_delay, self.connect)
-        else:
-            # TODO connect must not call failure callback from constructor. The
-            # current behavior is error-prone, because the user code may get a
-            # callback upon socket connection failure before user's other state
-            # may be sufficiently initialized. Constructors must either succeed
-            # or raise an exception. To be forward-compatible with failure
-            # reporting from fully non-blocking connection establishment,
-            # connect() should set INIT state and schedule a 0-second timer to
-            # continue the rest of the logic in a private method. The private
-            # method should use itself instead of connect() as the callback for
-            # scheduling retries.
+        # Schedule a timer callback to start the actual connection logic from
+        # event loop's context, thus avoiding error callbacks in the context of
+        # the caller, which could be the constructor.
+        self._connection_attempt_timer = self.add_timeout(
```

Comment:
This change also makes it so that there is a single code path for connection-establishment logic and the corresponding success/error-reporting logic, regardless of whether it's the initial connection attempt or a reconnection attempt.